int32 n_obstacles
ObstacleP[] obstacles